server/services/ai.ts

```diff
@@ -36,6 +188,780 @@ export default class AIService {
+            player.aiState.invasionsInProgress = player.aiState.invasionsInProgress.filter(invasion => invasion.arrivalTick > game.state.tick);
+        }
+    }
+
+    _createContext(game: Game, player: Player, aiState: AiState): Context {
+        const playerStars = this.starService.listStarsOwnedByPlayer(game.galaxy.stars, player._id);
+
+        const playerId = player._id.toString();
+
+        const starsById = new Map<string, Star>()
+
+        for (const star of game.galaxy.stars) {
+            starsById.set(star._id.toString(), star);
+        }
+
+        const traversableStars = game.galaxy.stars.filter(star => !star.ownedByPlayerId || star.ownedByPlayerId.toString() === playerId);
+        const allReachableFromPlayerStars = this._computeStarGraph(game, player, playerStars, game.galaxy.stars, this._getHyperspaceRange(game, player));
+        const allCanReachPlayerStars = this._computeStarGraph(game, player, game.galaxy.stars, playerStars, this._getHyperspaceRange(game, player));
+        const freelyReachableFromPlayerStars = this._computeStarGraph(game, player, playerStars, traversableStars, this._getHyperspaceRange(game, player));
+        const reachablePlayerStars = this._computeStarGraph(game, player, playerStars, playerStars, this._getHyperspaceRange(game, player));
+        const freelyReachableStars = this._computeStarGraph(game, player, traversableStars, traversableStars, this._getHyperspaceRange(game, player));
+        const starsInGlobalRange = this._computeStarGraph(game, player, playerStars, game.galaxy.stars, this._getGlobalHighestHyperspaceRange(game));
+        const borderStars = new Set<string>();
+        for (const [from, reachables] of starsInGlobalRange) {
+            for (const reachableId of reachables) {
+                const reachable = starsById.get(reachableId)!;
+                if (!reachable.ownedByPlayerId || reachable.ownedByPlayerId.toString() !== playerId) {
+                    borderStars.add(from);
+                }
+            }
+        }
+
+        const playerCarriers = this.carrierService.listCarriersOwnedByPlayer(game.galaxy.carriers, player._id);
+
+        const carriersOrbiting = new Map<string, Carrier[]>();
+        for (const carrier of game.galaxy.carriers) {
+            if ((!carrier.waypoints || carrier.waypoints.length === 0) && carrier.orbiting) {
+                const carriersInOrbit = getOrInsert(carriersOrbiting, carrier.orbiting.toString(), () => []);
+                carriersInOrbit.push(carrier);
+            }
+        }
+
+        const carriersById = new Map<string, Carrier>();
+        for (const carrier of game.galaxy.carriers) {
+            carriersById.set(carrier._id.toString(), carrier);
+        }
+
+        const incomingCarriers = game.galaxy.carriers
+            .filter(carrier => carrier.ownedByPlayerId!.toString() !== playerId)
+            .map(carrier => {
+                const waypoint = carrier.waypoints.find(wp => {
```

Comment:
Is this correct?

This looks like it checks if the carrier has **any** waypoint with the destination being owned by the AI player, shouldn't this be checking the current **in transit** waypoint?

Reply:
It's correct, but gives the AI a rather extreme advantage by being able to see the carrier before it's launched. I've changed it so it only takes the next waypoint into account.